src/native/ohos.rs

```diff
@@ -0,0 +1,578 @@
+                self.event_handler.key_up_event(
+                    self.context.with_display(&mut self.display),
+                    keycode,
+                    Default::default(),
+                );
+            }
+            Message::Pause => self
+                .event_handler
+                .window_minimized_event(self.context.with_display(&mut self.display)),
+            Message::Resume => {
+                self.event_handler
+                    .window_restored_event(self.context.with_display(&mut self.display))
+            }
+            Message::Destroy => {
+                self.quit = true;
+            }
+        }
+    }
+
+    fn frame(&mut self) {
+        self.event_handler
+            .update(self.context.with_display(&mut self.display));
+
+        if self.surface.is_null() == false {
+            self.event_handler
+                .draw(self.context.with_display(&mut self.display));
+
+            unsafe {
+                (self.libegl.eglSwapBuffers.unwrap())(self.egl_display, self.surface);
+            }
+        }
+    }
+
+}
+fn register_xcomponent_callbacks(xcomponent: &XComponent) -> napi_ohos::Result<()> {
+    let native_xcomponent = xcomponent.raw();
+    let res = unsafe {
+        OH_NativeXComponent_RegisterKeyEventCallback(native_xcomponent, Some(on_dispatch_key_event))
+    };
+    if res != 0 {
+        hilog_error!("Failed to register key event callbacks");
+    } else {
+        hilog_info!("Registered key event callbacks successfully");
+    }
+
+    Ok(())
+}
+
+pub unsafe extern "C" fn on_dispatch_key_event(
+    xcomponent: *mut OH_NativeXComponent,
+    window: *mut std::os::raw::c_void,
+)
+{
+    let mut event = std::ptr::null_mut();
+    let ret = OH_NativeXComponent_GetKeyEvent(xcomponent, &mut event);
+    assert!(ret == 0, "Get key event failed");
+
+    let mut action = 0;
+    let ret = OH_NativeXComponent_GetKeyEventAction(event, &mut action);
+    assert!(ret == 0, "Get key event action failed");
+
+    let mut code = 0;
+    let ret = OH_NativeXComponent_GetKeyEventCode(event, &mut code);
+    assert!(ret == 0, "Get key event code failed");
+
+    let keycode = keycodes::translate_keycode(code);
+    if action == 0 {
+        send_message(Message::KeyDown { keycode });
+    }
+    else if action == 1{
+        send_message(Message::KeyUp { keycode });
+    }
+}
+
+pub unsafe fn run<F>(conf: crate::conf::Conf, f: F)
+where
+    F: 'static + FnOnce(&mut crate::Context) -> Box<dyn EventHandler>,
+{
+    let env = OHOS_ENV.as_ref().expect("OHOS_ENV is not initialized");
+    let exports = OHOS_EXPORTS.as_ref().expect("OHOS_EXPORTS is not initialized");
+    let xcomponent = XComponent::init(*env, *exports).expect("Failed to initialize XComponent");
+    use std::panic;
+    panic::set_hook(Box::new(|info|{
+        hilog_fatal!(info)
+    }));
+    let _ = register_xcomponent_callbacks(&xcomponent);
+    struct SendHack<F>(F);
+    unsafe impl<F> Send for SendHack<F> {}
```

Comment:
Move this to top-level, rename to `AssumeSend` and add proper comments on why this should be Send